### PR TITLE
prometheusRule: move prometheus alert rules to values in order to be …

### PR DESCRIPTION
**Pull request by francoisminaud**
prometheusRule: move prometheus alert rules to values in order to be customizable

Signed-off-by: Francois Minaud <francois.minaud@gmail.com>

# Which chart

**Fluentd-elasticsearch**

# What this PR does / why we need it

So far the prometheus rules for alerting are hard coded in prometheusrule.yaml.
Our issue is that we want to customize some alerts. Indeed in our cluster we get some fluent retries quite often but do not want to trigger alerts after the first retry.

While the rules are hardcoded we cannot modify them. Moving them to the values file can be used to modify/add some.

# Which issue this PR fixes

*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

- fixes #

# Special notes for your reviewer

I am not a Helm expert. In order to get an accurate .Release.Name value in the values files I had to add another function in the template _helper.tpl. A more elegant solution might be possible.

# Checklist

- [x] [DCO](https://github.com/kokuwaio/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [ ] All variables are documented in the charts README

### CODE REVIEW

File: charts/fluentd-elasticsearch/templates/prometheusrule.yaml [hunk 1]
```diff
@@ -11,77 +11,9 @@ metadata:
     {{- end }}
   namespace: {{ .Values.prometheusRule.prometheusNamespace }}
 spec:
+  {{- with .Values.prometheusRule.rules }}
```
Comment (monotek) [inline charts/fluentd-elasticsearch/templates/prometheusrule.yaml:14]: imho this will not work as there are variables used inside the rule.
tpl function could be used instead.

https://helm.sh/docs/howto/charts_tips_and_tricks/#using-the-tpl-function

File: charts/fluentd-elasticsearch/Chart.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 apiVersion: v2
 name: fluentd-elasticsearch
-version: 11.12.0
+version: 11.12.1
```
Comment (monotek) [inline charts/fluentd-elasticsearch/Chart.yaml:3]: as this is a new feature minor version should be raised.

File: charts/fluentd-elasticsearch/templates/_helpers.tpl [hunk 3]
```diff
@@ -23,6 +23,13 @@ If release name contains chart name it will be used as a full name.
 {{- end }}
 {{- end }}
 
+{{/*
+Returns the chart Release Name.
+*/}}
+{{- define "fluentd-elasticsearch.releaseName" -}}
+{{- .Release.Name | trunc 63 | trimSuffix "-" }}
+{{- end }}
+
```
Comment (monotek) [inline charts/fluentd-elasticsearch/templates/_helpers.tpl:32]: why is this needed?
"Release.Name" is already available. 
See: https://helm.sh/docs/chart_template_guide/builtin_objects/

**Review by monotek — CHANGES_REQUESTED** [review]
please aslo sign your commit, to get the dco check passed.

File: charts/fluentd-elasticsearch/Chart.yaml (hunk 2 above)
Comment (francoisminaud) [reply charts/fluentd-elasticsearch/Chart.yaml]: I indeed raised the minor version instead.

File: charts/fluentd-elasticsearch/templates/_helpers.tpl (hunk 3 above)
Comment (francoisminaud) [reply charts/fluentd-elasticsearch/templates/_helpers.tpl:32]: I've had to use it because Release.Name is not directly available in the values.yaml file. It is only available in templating function.

The alternative was using the ["fluentd-elasticsearch.fullname"](https://github.com/kokuwaio/helm-charts/blob/main/charts/fluentd-elasticsearch/templates/_helpers.tpl#L13) alas the rendering of the PrometheusRule file was not the same as previously when .Release.Name was directly defined in the prometheusrule.yaml file.

The main issue is this rule: https://github.com/kokuwaio/helm-charts/blob/main/charts/fluentd-elasticsearch/templates/prometheusrule.yaml#L58
When using the templated fluentd-elasticsearch.fullname function, the rendered value was not the Release Name, this is why I had to include this function.

File: charts/fluentd-elasticsearch/templates/prometheusrule.yaml (hunk 1 above)
Comment (francoisminaud) [reply charts/fluentd-elasticsearch/templates/prometheusrule.yaml:14]: I have taken the bitnami redis chart (which we are using as example) and the rules are rendered as expected 
https://github.com/bitnami/charts/blob/master/bitnami/redis/templates/prometheusrule.yaml#L25

File: charts/fluentd-elasticsearch/templates/_helpers.tpl (hunk 3 above)
Comment (monotek) [reply charts/fluentd-elasticsearch/templates/_helpers.tpl]: Probably because of the scope. 
See. https://helm.sh/docs/chart_template_guide/variables/

But in tpl it works. I've changed it.

**Review by monotek — DISMISSED** [review]

**Comment by francoisminaud** [comment]
I modified the _helper template and removed the new Release.Name and used template in the values.yaml file.
Hope that looks better for you.

**Comment by monotek** [comment]
I've already did. Seems you overwrote my changes.

File: charts/fluentd-elasticsearch/templates/prometheusrule.yaml (hunk 1 above)
Comment (monotek) [inline charts/fluentd-elasticsearch/templates/prometheusrule.yaml:14]: thats superfluous. If we don't want the rules to be used we can just use .Values.prometheusRule.enabled=false

File: charts/fluentd-elasticsearch/values.yaml [hunk 4]
```diff
@@ -299,6 +299,73 @@ prometheusRule:
...
+  - alert: FluentdOutputError
+    expr: increase(fluentd_output_status_num_errors[10m]) > 0
+    for: 1s
+    labels:
+      service: fluentd
+      severity: warning
+    annotations:
+      description: Fluentd output error count is {{ "{{ $value }}" }} for the last 10 minutes
+      summary: There have been Fluentd output error(s) for the last 10 minutes
   #  role: alert-rules
```
Comment (monotek) [inline charts/fluentd-elasticsearch/values.yaml:369]: useless commnet

File: charts/fluentd-elasticsearch/templates/prometheusrule.yaml [hunk 5]
```diff
@@ -11,77 +11,9 @@ metadata:
...
-      expr: increase(fluentd_output_status_num_errors[10m]) > 0
-      for: 1s
-      labels:
-        service: fluentd
-        severity: warning
-      annotations:
-        description: Fluentd output error count is {{ "{{ $value }}" }} for the last 10 minutes
-        summary: There have been Fluentd output error(s) for the last 10 minutes
-
+    rules: {{- tpl (toYaml .) $ | nindent 4 }}
```
Comment (monotek) [inline charts/fluentd-elasticsearch/templates/prometheusrule.yaml:17]:
```suggestion
    rules: {{- tpl (toYaml .Values.prometheusRule.rules) $ | nindent 4 }}
```

**Review by monotek — CHANGES_REQUESTED** [review]

File: charts/fluentd-elasticsearch/templates/prometheusrule.yaml (hunk 5 above)
Comment (francoisminaud) [reply charts/fluentd-elasticsearch/templates/prometheusrule.yaml]: Sorry about the mess, I pushed the suggested changes.

**Review by monotek — APPROVED** [review]

**Comment by francoisminaud** [comment]
Thanks!
